sympy/sets/sets.py

```diff
@@ -735,6 +768,35 @@ def is_iterable(self):
         """
         return all(set.is_iterable for set in self.sets)
 
+    @property
+    def is_finite_set(self):
+        """A property to check if the set has finite number of elements.
+
+        Examples
+        ========
+
+        >>> from sympy.sets.sets import FiniteSet, ProductSet
+
+        >>> a = FiniteSet(1, 2)
+        >>> b = ProductSet(a, a)
+        >>> b.is_finite_set
+        True
+
+        Notes
+        =====
+
+        Only the lemma that the finite Cartesian product of finite sets is a
+        finite set is implemented.
+
+        References
+        ==========
+
+        .. [1] https://proofwiki.org/wiki/Product_of_Finite_Sets_is_Finite
+        """
+        if all(arg.is_finite_set for arg in self.args):
+            return True
+        return False
```

Comment:
I guess that some sets could have `is_finite_set` as None e.g.:
```julia
In [6]: a, x = symbols('a, x')

In [7]: C = ConditionSet(x, Eq(a*x,0), S.Reals)                                                                                                                                                                    

In [8]: C
Out[8]: {x | x ∊ ℝ ∧ a⋅x = 0}

In [9]: C.subs(a, 1) # should simplify to {0}
Out[9]: {x | x ∊ ℝ ∧ x = 0}

In [10]: C.subs(a, 0)
Out[10]: ℝ
```
If this was ternary logic then it should be something like:
```
fuzzy_and(arg.is_finite_set for arg in self.args)
```
Might also need to check if any arg is the empty set.